### PR TITLE
Enable merge option for datasets (phase 1)

### PR DESCRIPTION
**Pull request by Sigmonia**
#### Rationale
Spec: https://docs.google.com/document/d/110fl8qRIk1o9lEgFq8RKFWNW5vo_cYCiWkE1doFq214/
Enable the Merge option for bulk updating Datasets

TODO: (Phase 2) Improve Audit Logging

#### Changes
* Enabled UI option for the Dataset bulk insert page and form
* Passed importOption and auditLogBehavior parameters through to the DataIterator context

### CODE REVIEW

File: study/src/org/labkey/study/controllers/DatasetController.java [hunk 1]
```diff
@@ -197,7 +175,7 @@ public ModelAndView getView(DatasetAuditHistoryForm form, BindException errors)
             @Override
             protected void renderInternal(Object model, PrintWriter out)
             {
-                out.write("<p>No current record found</p>");
+                out.write("<p>No additional details recorded</p>");
```
Comment (labkey-susanh) [inline study/src/org/labkey/study/controllers/DatasetController.java:178]: Any tests that need updating as a result of this text change?

File: study/src/org/labkey/study/controllers/DatasetController.java [hunk 2]
```diff
@@ -137,32 +137,6 @@ public ModelAndView getView(DatasetAuditHistoryForm form, BindException errors)
             if (newRecord != null)
             {
                 newData = DatasetAuditProvider.decodeFromDataMap(newRecord);
-                String lsid = newData.get("lsid");
```
Comment (labkey-susanh) [inline study/src/org/labkey/study/controllers/DatasetController.java:140]: Does the removal of this check for the lsid mean we'll never see the details?  Or is the case when we have the details to show handled elsewhere?

File: study/src/org/labkey/study/controllers/DatasetController.java (hunk 1 above)
Comment (Sigmonia) [reply study/src/org/labkey/study/controllers/DatasetController.java:178]: Not that i could find.

File: study/src/org/labkey/study/controllers/DatasetController.java (hunk 2 above)
Comment (Sigmonia) [reply study/src/org/labkey/study/controllers/DatasetController.java:140]: This removes the duplicated block seen on the new inserts. The block starting at Line 172 (old): 146 (change) should continue to show the new row or diff as appropriate.

Comment (labkey-susanh) [reply study/src/org/labkey/study/controllers/DatasetController.java:140]: Got it.

**Review by labkey-susanh — APPROVED** [review]

File: study/src/org/labkey/study/model/DatasetDefinition.java [hunk 3]
```diff
@@ -1576,15 +1581,43 @@ public TableInfo getSchemaTableInfo()
             return m;
         }
 
+        public PHI getMaxContainedPhi()
+        {
+            return _maxContainedPhi;
+        }
+
         @Override
-        public CaseInsensitiveHashSet skipProperties()
```
Comment (Sigmonia) [inline study/src/org/labkey/study/model/DatasetDefinition.java:1580]: Removed as it matches the underlying method

File: study/src/org/labkey/study/model/DatasetDefinition.java [hunk 4]
```diff
@@ -1576,15 +1581,43 @@ public TableInfo getSchemaTableInfo()
             return m;
         }
 
+        public PHI getMaxContainedPhi()
```
Comment (Sigmonia) [inline study/src/org/labkey/study/model/DatasetDefinition.java:1584]: Not changed, just a bad diff.

File: study/src/org/labkey/study/model/DatasetDefinition.java [hunk 5]
```diff
@@ -1576,15 +1581,43 @@ public TableInfo getSchemaTableInfo()
...
+        public void addAuditEvent(User user, Container container, AuditBehaviorType auditBehavior, @Nullable String userComment, QueryService.AuditAction auditAction, Map<String, Object> parameters)
         {
-            return _maxContainedPhi;
+            Map<String,Object> oldRow = null;
+            //Leaving this block as reference for getting update diff
+//            if (auditAction == QueryService.AuditAction.MERGE)
+//            {
+//                String lsid = String.valueOf(parameters.get("lsid"));
+//                oldRow = DatasetDefinition.this.getDatasetRow(user, lsid);
+//            }
```
Comment (labkey-susanh) [inline study/src/org/labkey/study/model/DatasetDefinition.java]: This code is in the other branch (https://github.com/LabKey/platform/tree/20.11_fb_dataset_merge_auditing),  so I don't think this commented out block is necessary.

**Review by labkey-susanh — APPROVED** [review]
As we discussed, let's add the check on auditBehavior in the two addAuditLog methods just in case the methods are called not from the DataIterator.  

Otherwise, looks good.

**Pull request merged**